src/libs/actions/SignInRedirect.js

```diff
@@ -24,8 +25,10 @@ function clearStorageAndRedirect(errorMessage) {
     const preferredLocale = currentPreferredLocale;
 
     // Clearing storage discards the authToken. This causes a redirect to the SignIn screen
+    waitForOnyxToClear.setOnyxClearing();
```

Comment:
OK, I think it's mostly the name of this which is tripping me up, but there are a few more things...

- It seems like a better convention that would make sense is something more like `Session.signOut().then()` since that's what it feels like we're actually waiting for (this assumes I'm understanding the code properly, which I'm not 100% sure that I am yet).
- If `Onyx.clear()` returns a promise, why aren't we just using that promise? (this would mean adding a `return` to `clearStorageAndRedirect` and `redirectToSignIn`)
- `clearStorageAndRedirect()` doesn't actually do any redirecting, so this is a bad name and really throwing me for a loop. I can't find where the redirect is happening.

Am I correctly understanding what is really happening is that `LoginWithShortLivedTokenPage.js` is waiting for `clearStorageAndRedirect()` to be done, which is actually triggered from `LogOutPreviousUserPage.js`? If that's the case, then I think it is more clear that `clear()` should be finished _BEFORE_ the redirect to `LoginWithShortLivedTokenPage`. That way, `LoginWithShortLivedTokenPage` doesn't have to worry at all about if `clear()` was done or not.

Reply:
Maybe you figured these out already, but I'll respond for clarity.
> * Why is an action file managing the internal state of Onyx?

It is only creating a promise that waits for Onyx to finish clearing
> * Shouldn't Onyx be in charge of managing its own internal state?

N/A
> * If something is waiting for this clear to be done, then why don't I see a `promise` anywhere?

The promise lives in `waitForOnyxToClear` because multiple files need to access that promise.

Reply:
> OK, I think it's mostly the name of this which is tripping me up, but there are a few more things...

How about `waitForOnyxToClear.setIsOnyxClearing();`? That might make it more clear that we are setting our own state rather than Onyx internal state.
> * It seems like a better convention that would make sense is something more like `Session.signOut().then()` since that's what it feels like we're actually waiting for (this assumes I'm understanding the code properly, which I'm not 100% sure that I am yet).

Ya I think that could work, if we made `Session.signOut` return the `Onyx.clear` promise.
> * If `Onyx.clear()` returns a promise, why aren't we just using that promise? (this would mean adding a `return` to `clearStorageAndRedirect` and `redirectToSignIn`)

Yep, we could do that.
> * `clearStorageAndRedirect()` doesn't actually do any redirecting, so this is a bad name and really throwing me for a loop. I can't find where the redirect is happening.

That confused me initially too. How about `clearStorage` or `resetStorage` or `clearStorageAndReset`? Also, the "redirect" happens because the session gets cleared out, which mounts the `PublicScreens` instead, which renders `LogInWithShortLivedTokenPage` on the `TRANSITION_FROM_OLD_DOT` route. https://github.com/Expensify/App/blob/f3b265d4a1785fe8fa3114869eacc886186ea260/src/libs/Navigation/AppNavigator/PublicScreens.js#L19-L23

> Am I correctly understanding what is really happening is that `LoginWithShortLivedTokenPage.js` is waiting for `clearStorageAndRedirect()` to be done, which is actually triggered from `LogOutPreviousUserPage.js`? 

Yep.
> If that's the case, then I think it is more clear that `clear()` should be finished _BEFORE_ the redirect to `LoginWithShortLivedTokenPage`. That way, `LoginWithShortLivedTokenPage` doesn't have to worry at all about if `clear()` was done or not.

Yeah that would be nice, but we would have to change the way the "redirect" is handled. When we clear Onyx we will reset the session at some point which will cause the redirect, but this will happen before the clear finishes. If we can fix the Onyx clear behavior with this PR then we wouldn't have to do any of this waiting for onyx stuff.

I don't think we want to change the "redirect" behavior because it's good to have one route that we can link to which will handle all 3 cases where a user is signed in, out, our signed in to a different account.

Reply:
>It seems like a better convention that would make sense is something more like Session.signOut().then()
>I think it is more clear that clear() should be finished BEFORE the redirect to LoginWithShortLivedTokenPage. That way, LoginWithShortLivedTokenPage doesn't have to worry at all about if clear() was done or not.

Totally following this train of thought, but just wanted to remind that it's the removal of the `authToken` that will move us to `LoginWithShortLivedTokenPage`. That's why waiting for the sign out to happen doesn't quite work. The navigation isn't programmatic as it depends on the `authToken` state.

>If Onyx.clear() returns a promise, why aren't we just using that promise? (this would mean adding a return to clearStorageAndRedirect and redirectToSignIn)
> Shouldn't Onyx be in charge of managing its own internal state

This made me wonder about whether maybe a callback would be better than a promise and if we could make this a feature of `Onyx`.

```js
if (!Onyx.isClearing) {
    doSomething();
    return;
}

Onyx.onClear(doSomething);
```

Reply:
Hmm, now I'm not sure about renaming `clearStorageAndRedirect` since this is effectively what it does even though there isn't a programmatic redirect.

I renamed some of the `waitForOnyxToClear` methods. @tgolen @marcaaron what additional changes should I make based on this comment thread if any?

Reply:
Ah, OK, I totally forgot about the clearing of the `authToken` is what causes the redirect. I think there are some things that could make it more clear:

1. Adding a comment in the code directly explaining how and why the redirect occurs
2. Rename it to something more like `resetSessionToLogUserOut`

I'm still a little bit confused about the race condition with the storage being cleared. If it's the storage being cleared which triggers the redirect, then how is it at all possible to land on `LoginWithShortLivedTokenPage` before the clear is done? I can imagine it would happen where there is something like... OK, 20 keys need to be reset, and the session key is reset first, so the redirect happens, and the other 19 keys are still being reset in the meantime. However, even in that case, why does it matter to `LoginWithShortLivedTokenPage`? As long as the session is gone, isn't that all it cares about?

Reply:
> This made me wonder about whether maybe a callback would be better than a promise and if we could make this a feature of Onyx.

I wanted to address this quickly. I think if there was one downfall I have with promises is that they can't be resolved more than once. I think this is why you are suggesting a callback instead?